### ISSUE TITLE
TRF034: false positive on every modular file whose layer inherits an imported layer (49/75 findings wrong)

### ISSUE THREAD

**Issue by tarekziade** [huggingface/transformers-mlinter#36]
## Summary

`TRF034` walks only the **local** base chain, so it fires on any `modular_*.py` whose layer class
inherits an imported layer — the standard modular pattern. Reported by @Rocketknight1 on
[huggingface/transformers#46419](https://github.com/huggingface/transformers/pull/46419#issuecomment-5315552928)
(ESMC port), where it was correctly refused as a linter bug.

## Reproducer

From that PR's head (`port-esmc-esmfold2`):

```
$ mlinter src/transformers/models/esmc/modular_esmc.py
src/transformers/models/esmc/modular_esmc.py:L242: TRF034: `EsmcLayer` is stacked in an `nn.ModuleList`
but does not subclass `GradientCheckpointingLayer` [...]
```

but the class is fine:

```python
# modular_esmc.py:190
class EsmcLayer(LlamaDecoderLayer):        # LlamaDecoderLayer -> GradientCheckpointingLayer
# modeling_esmc.py:259  (generated, this is what actually runs)
class EsmcLayer(GradientCheckpointingLayer):
```

## Root cause

`_subclasses_gradient_checkpointing_layer` (`mlinter/trf034.py:17`) resolves bases from
`_collect_class_bases(tree)`, i.e. classes defined **in the linted file only**. A modular file gets
its bases by import (`from ..llama.modeling_llama import LlamaDecoderLayer`), so the walk dead-ends
on an unknown name and the rule reports.

This is structural, not incidental: the driver deliberately skips generated files
(`mlinter.py::_is_generated_file`), so `modular_*.py` is the *only* file linted — precisely the one
where the base chain is invisible. Compare `_inherits_pretrained_model`, which matches a name
*suffix* (`*PreTrainedModel`) and therefore survives cross-file inheritance; TRF034 matches an exact
class name that never appears locally.

## Scale (transformers `main`, 2026-08-18)

| | count |
|---|---|
| TRF034 findings under `src/transformers/models` | 131 |
| ... on `modular_*.py` | 75 |
| ... **false positives** (generated sibling does reach `GradientCheckpointingLayer`) | **49** |

So ~37% of the rule's output is wrong. Affected models include `aria`, `cwm`, `emu3`, `exaone4`,
`gemma3n`, `granite*`, `helium`, `janus`, `olmo{,2,3,e}`, `qwen2_moe`, `qwen3_next`,
`qwen3_omni_moe`, `qwen3_vl`, `starcoder2`, `t5gemma2`, `xlm_roberta_xl`. It stays mostly invisible
only because CI runs `--changed-only` — it surfaces whenever someone touches one of those files, or
adds a new modular model, as ESMC did. The 10 `allowlist_models` entries are not masking these
(re-checked with the allowlist and cutoff neutralised: their 3 findings are genuine).

## Proposed fix

Resolve bases across relative imports, with a **tri-state** result:

- reaches `GradientCheckpointingLayer` → no violation;
- chain fully resolved and terminates at `nn.Module` / `object` → report;
- some base unresolvable (external package, standalone model repo) → **fail open**, no violation.

Resolution is mechanical: `ast.ImportFrom(level=n, module="llama.modeling_llama")` from
`.../models/esmc/modular_esmc.py` → walk up `n-1` dirs → `.../models/llama/modeling_llama.py`; parse,
continue the chain, cache per file, cap depth. There is precedent for cross-file lookups in
`trf022.py::_model_dir_class_names`.

**Validated on a prototype** over all 134 findings (allowlist + cutoff neutralised):

```
findings=134  suppressed_correctly=49  still_reported_correctly=85  still_FP=0  lost_true_positives=0  unknown=0
```

Worth noting the cheaper heuristic ("base isn't local → skip") is *not* good enough: it silently
drops 4 real violations where a modular class inherits an `RTDetr*` layer that is itself a plain
`nn.Module` (`rt_detr_v2`, `d_fine`). Import resolution keeps them.

Also needs updating once fixed: the `what_it_does` text in `rules.toml` ("through its local base
chain") and `docs/rules/trf034.md`, plus a regression test for a modular file inheriting an imported
`GradientCheckpointingLayer` subclass.
